src/care/host_device_map.h

```diff
@@ -90,34 +94,52 @@ namespace care {
         // prepare for lookups, update our size from the map size
         void sort() {
            *m_size = m_map->size();
+           *m_iterator = m_map->begin();
+           *m_next_iterator_index = 0;
         } 
         
 
         // free any heap data
         void free() {
            delete m_map;
            delete m_size;
+           delete m_iterator;
+           delete m_next_iterator_index;
         }
 
         // return the number of inserted elements
         int size() const { return *m_size; }
 
         // clear any added elements
-        void clear() { m_map->clear();}
+        void clear() { 
+           m_map->clear();
+           *m_size = 0;
+           *m_iterator = m_map->begin();
+           *m_next_iterator_index = 0;
+        }
         
         // preallocate buffers for adding up to size elements
         void reserve(int max_size) {
            m_max_size = max_size;
         }
 
         // iteration - meant to only be called by the CARE_MAP_LOOP macros
-        inline typename std::map<key_type, mapped_type>::iterator begin () const { return m_map->begin(); }
-        inline typename std::map<key_type, mapped_type>::iterator end () const { return m_map->end(); }
-        
+        inline typename std::map<key_type, mapped_type>::iterator iterator_at(int index) const {
+           if (index != *m_next_iterator_index) {
+              *m_iterator = m_map->begin();
+              for (*m_next_iterator_index = 0; *m_next_iterator_index < index ; ++(*m_next_iterator_index)) {
+                 ++(*m_iterator);
+              }
+           }
+           ++(*m_next_iterator_index);
+           return (*m_iterator)++;
+        }
 
       private:
          // we do a heap allocated map to ensure no deep copies occur during lambda capture
          std::map<key_type, mapped_type> * m_map = nullptr;
+         mutable typename std::map<key_type, mapped_type>::iterator * m_iterator = nullptr;
```

Comment:
you shouldn't need the mutables since the pointer value itself isn't changing in a lambda context

Reply:
Ah, yes.  I think I was missing a dereference in an earlier version (which caused the compiler to complain about const-ness).